src/mlpack/methods/ann/layer/multihead_attention_impl.hpp

```diff
@@ -0,0 +1,288 @@
+  valueBias = arma::mat(weights.memptr()
+      + (4 * embedDim + 2) * embedDim, embedDim, 1, false, false).zeros();
+  outBias = arma::mat(weights.memptr()
+      + (4 * embedDim + 3) * embedDim, embedDim, 1, false, false).zeros();
+  XavierInitialization().Initialize(queryWt, embedDim, embedDim);
+  XavierInitialization().Initialize(keyWt, embedDim, embedDim);
+  XavierInitialization().Initialize(valueWt, embedDim, embedDim);
+  XavierInitialization().Initialize(outWt, embedDim, embedDim);
+}
+
+template <typename InputDataType, typename OutputDataType,
+          typename RegularizerType>
+template <typename eT>
+void MultiheadAttention<InputDataType, OutputDataType, RegularizerType>::
+Forward(const arma::Mat<eT>& input, arma::Mat<eT>& output)
+{
+  Log::Assert(input.n_rows == embedDim * (tLen + 2 * sLen));
+  const size_t bsz = input.n_cols;
+  output.set_size(embedDim * tLen, bsz);
+
+  arma::cube query(const_cast<arma::Mat<eT>&>(input).memptr(),
+      embedDim, tLen, bsz, false, false);
+  arma::cube key(const_cast<arma::Mat<eT>&>(input).memptr()
+      + query.n_elem, embedDim, sLen, bsz, false, false);
+  arma::cube value(const_cast<arma::Mat<eT>&>(input).memptr()
+      + query.n_elem + key.n_elem, embedDim, sLen, bsz, false, false);
+
+  for (size_t i = 0; i < bsz; ++i)
+  {
+    query.slice(i) = queryWt * query.slice(i) + arma::repmat(queryBias,1, tLen);
+    key.slice(i) = keyWt * key.slice(i) + arma::repmat(keyBias, 1, sLen);
+    value.slice(i) = valueWt * key.slice(i) + arma::repmat(valueBias, 1, sLen);
+  }
+  query /= std::sqrt(headDim);
+
+  if (!attnMask2d.is_empty())
+  {
+    if (attnMask2d.n_rows != sLen || attnMask2d.n_cols != tLen)
+      Log::Fatal << "The size of the 2D `attn_mask` is not correct."
+                 << std::endl;
+  }
+  else if (!attnMask3d.is_empty())
+  {
+    if (attnMask3d.n_rows != sLen || attnMask3d.n_cols != tLen ||
+        attnMask3d.n_slices != bsz * numHeads)
+      Log::Fatal << "The size of the 3D `attn_mask` is not correct."
+                 << std::endl;
+  }
+
+  if (!keyPaddingMask.is_empty())
+  {
+    if (keyPaddingMask.n_rows != sLen || keyPaddingMask.n_cols != 1)
+        Log::Fatal << "The size of the `keyPaddingMask` is not correct."
+                   << std::endl;
+  }
+
+  qProj = arma::cube(query.memptr(), headDim, tLen, numHeads * bsz, 0, 0);
+  kProj = arma::cube(key.memptr(), headDim, sLen, numHeads * bsz, 0, 0);
+  vProj = arma::cube(value.memptr(), headDim, sLen, numHeads * bsz, 0, 0);
+  arma::cube Wt3d(sLen, tLen, numHeads * bsz);
+
+  for (size_t i = 0; i < numHeads * bsz; ++i)
+  {
+    Wt3d.slice(i) = kProj.slice(i).t() * qProj.slice(i);
+  }
+
+  if (!attnMask2d.is_empty())
+    Wt3d.each_slice() += attnMask2d;
+  else if (!attnMask3d.is_empty())
+    Wt3d += attnMask3d;
+
+  if (!keyPaddingMask.is_empty())
+    for (size_t i = 0; i < Wt3d.n_slices; ++i)
+    {
+      Wt3d.slice(i) += arma::repmat(keyPaddingMask, 1, tLen);
+    }
```

Comment:
Looks like you do the same thing using three different ways. I was wondering if it's possible to simplify it a bit.

Reply:
Here, `Wt3d` is a 3d tensor. So, when applying 2d mask, I will have to iterate to each slice and apply the mask. Now when the mask is 3d, I need not do that, I can simply add two cubes.
Yes, for `keyPaddingMask`, it can be simplified by just using `.each_slice()` method. Thanks for pointing it out.